File: Strings/Space20/main.cpp
/*
 * Given a string, write a function to replace all spaces in a string with '%20'.
 * It is given that the string has sufficient space at the end to hold the
 * additional characters.
 */
#include <iostream>
#include <vector>
#include <string>

using namespace std;

string REPLACE_WITH = "%20";

vector<int> getAllOccurrences(string s, string term)
{
    vector<int> indexes;
    int idx = s.find(term);
    while (idx != -1)
    {
        indexes.push_back(idx);
        idx = s.find(term, idx + 1);
    }

    return indexes;
}

int getAdditionalLen(string s)
{
    vector<int> occurrences = getAllOccurrences(s, " ");
    return occurrences.size() * (REPLACE_WITH.size() - 1);
}

string resizeString(string s)
{
    int additionalLen = getAdditionalLen(s);
    s.resize(s.size() + additionalLen);

    return s;
}

string space20Quadratic(string s)
{
    int stringLen = s.size();
    s = resizeString(s);

    int diff = REPLACE_WITH.size() - 1;
    for (int i = 0; i < stringLen;)
    {
        if (s[i] == ' ')
        {
            int j;
            for (j = stringLen - 1; j > i; j--)
            {
                s[j + diff] = s[j];
            }

            for (j = 0; j < REPLACE_WITH.size(); j++)
            {
                s[i] = REPLACE_WITH[j];
                i++;
            }

            stringLen += diff;
        }
        else
        {
            i++;
        }
    }

    return s;
}

string space20Linear(string s)
{
    int stringLen = s.size();
    int additionalLen = getAdditionalLen(s);

    s = resizeString(s);
    int diff = REPLACE_WITH.size() - 1;
    for (int i = stringLen - 1; i >= 0; i--)
    {
        int moveTo = i + additionalLen;
        if (s[i] == ' ')
        {
            int stop = moveTo - diff;
            int j = moveTo;
            while (j >= stop)
            {
                s[j] = REPLACE_WITH[j - stop];
                j--;
            }

            additionalLen -= diff;
        }
        else
        {
            s[moveTo] = s[i];
        }
    }

    return s;
}

int main()
{
    string s = "hello world,  how are you?";
    cout << s << endl;
    cout << space20Quadratic(s) << endl;
    cout << space20Linear(s) << endl;

    return 0;
}

File: Strings/Space20/main.py
REPLACE_WITH = "%20"
SPACE_LEN = len(" ")


def get_all_occurrences(s: str, term: str):
    indexes = []
    idx = s.find(term)
    while idx != -1:
        indexes.append(idx)
        idx = s.find(term, idx + 1)

    return indexes


def get_additional_len(s: str):
    occurrences = get_all_occurrences(s, " ")
    return len(occurrences) * (len(REPLACE_WITH) - 1)


def resize_string(s: str):
    new_str = s[:]
    additional_len = get_additional_len(s)
    new_len = len(s) + additional_len

    return new_str.ljust(new_len)


def space20_quadratic(s: str):
    string_len = len(s)
    s = list(resize_string(s))

    diff = len(REPLACE_WITH) - SPACE_LEN
    i = 0
    while i < string_len:
        if s[i] == " ":
            for j, char in reversed(list(enumerate(s[i:string_len], start=i))):
                s[j + diff] = char

            for j, char in enumerate(REPLACE_WITH):
                s[i] = char
                i += 1

            string_len += diff
        else:
            i += 1

    return "".join(s)


def space20_linear(s: str):
    original_len = len(s)
    additional_len = get_additional_len(s)

    s = list(resize_string(s))
    diff = len(REPLACE_WITH) - SPACE_LEN
    for i, char in reversed(list(enumerate(s[:original_len]))):
        move_to = i + additional_len
        if char == " ":
            stop = move_to - diff
            j = move_to
            while j >= stop:
                s[j] = REPLACE_WITH[j - stop]
                j -= 1

            additional_len -= diff
        else:
            s[move_to] = s[i]

    return "".join(s)


def main():
    s = "hello world,  how are you?"
    print(s)
    print(space20_quadratic(s))
    print(space20_linear(s))


if __name__ == "__main__":
    main()
